### ISSUE TITLE
Figure out attribute binding syntaxes

### ISSUE THREAD

**Issue by justinbmeyer** [bitovi/canjs#1700]
In combination with #1699, this proposal seeks to create a more flexible, powerful, and easily understood attribute binding syntaxes for use in stache.

## Summary

| Type                 | Example  |
|-------------------------|--------------------------|
| event - viewModel | `<my-comp (child-event)="parentMethod()"/>` |
| event - DOM | `<my-comp ($click)="parentMethod()"/>` |
| 0 way - viewModel  | `<my-comp child-prop="foo"/>` |
| one way - viewModel - parent to child | `<my-comp {child-prop}="parentProp"/>` |
| one way - DOM - parent to element | `<my-comp {$el-prop}="parentProp"/>` |
| one way - reference - child to reference | `<my-comp {^child-prop}="*ref"/>`        |
| two way - viewModel | `<my-comp {(child-prop)}="parentProp"/>`  |
| two way - DOM | `<my-comp {($child-prop)}="parentProp"/>`  |
| two way - reference | `<my-comp {(child-prop)}="*ref"/>`        |
| two way - reference shorthand | `<my-comp *ref/>`        |

## event - viewModel

`(event-name)="StacheMethodExpression"`

Binds an event on the `can.viewModel` of whatever element it was placed on.

Example:

```
<modal (todo.created)="hideModal()" (cancelled)="hideModal()"/>
```

## event - DOM

`($event-name)="StacheMethodExpression"`

Bind an event on the element it was placed on.

Example:

```
<button ($click)="todo.save()">Save</button>
```

## 0 way - viewModel

`view-model-attr="value"`

Sets the `viewModelAttr` property to `"value"`.

## one way - viewModel - parent to child

The following looks up `task` in the scope, and sets it as `todo` in the `<modal>`'s viewModel:

```
<modal {todo}="task"/>
```

## one way - DOM - parent to element

Provides one-way binding to an element's property or attribute.  We will maintain a list of properties that will be set instead of setting the attribute.

```
<input {$checked}="complete" type='checkbox'/>
```

This is equivalent to:

 ```
<input {{#complete}}checked{{/complete}} type='checkbox'/>
```

While it's possible to do one-way binding to an element property or attribute with stache alone, two-way binding uses this for syntactic sugar. 

## one way - reference - child to reference

The following looks up `task` in the child viewModel, and sets it as `todo` in the current template's 
reference scope:

```
<modal {^task}="*todo"/>
```

## two way - viewModel

This two-way binds a viewModel property to something in the scope.  The following two-way binds `<component>`'s `childProp` to `parentProp` in the scope:

```html
<component  {(child-prop)}="parentProp" />
```

This is syntactic sugar for:

```html
<component  {childProp}="parentProp"
           (childProp)="~parentProp(%viewModel.childProp)" />
```

## two way - DOM

This two-way binds a element property to something in the scope.  The following two-way binds the `<input>`'s `checked` property to `complete` in the scope:

```html
<input {($checked)}="complete" type='checkbox' />
```

This is syntactic sugar for:

```html
<component  {$checked}="complete"
           (change)="~complete($element.checked)" />
```

A mapping of property names to which event should be listened to will be kept. 

## two way - reference

This two-way binds a element property to something in the scope.  The following two-way binds the `<component>`'s `selected.licensePlate` property to `selectedPlate` in the template references scope:

```html
<component {(selected.license-plate)}="*selectedPlate"/>
```

This is syntactic sugar for:

```html
<component  {selected.license-plate}="*selectedPlate"
           (selected.licensePlate)="~*selectedPlate(%viewModel.selected.licensePlate)" />
```

**Comment by matthewp** [comment]
I'm thinking the "into variable" use-case will be most common.  I almost think we should make it explicit if you want to place it on the parent context. What about:

```mustache
<component [^parent-prop]="{child}">
```

With the `^` meaning to add to the parent.  I don't like the `{{#}}` idea because it puts a tax on the most common need.

**Comment by matthewp** [comment]
Here's how Angular 2 does bindings for comparison:

## Bindings

All bindings use the [foo]="{bar}" syntax like us, but the opposite direction.  This binds to an input's value:

```html
<input type="text" [value]="{{scope.value}}">
```

And this would also bind to a custom component's value:

```html
<component [foo]="{{bar}}">
```

This is interesting because the fact that they bind to an element's properties means you can do things like:

```html
<div [inner-html]="{{scope.myHtml}}"></div>
```

## Reference

This is a way to reference a component within the view. So you can do:

```html
<input #myInput type="text">

<span> {{myInput.value}} </span>
```

**Comment by matthewp** [comment]
(read the [above comment](https://github.com/bitovi/canjs/issues/1700#issuecomment-104739103) first)

I like having a different syntax for reference binding.  This could solve our issue of needing a "local variable" syntax.  So you could do:

```html
<can-import from="my/child" #child />

{{child.state}}
```

`#child` is a promise, this will show "pending" and then "resolved" (or "rejected")

I think you should also be able to do:

```html
<can-input from="my/child" #child="{value}" />

{{child}}
```

Here we are binding to the promise's `value`.

**Comment by matthewp** [comment]
@justinbmeyer what was the use case for one-way binding?  And how would that work in practice?

**Comment by justinbmeyer** [comment]
```
TWO WAY:        <input value="{name}"/>
ONE WAY p->c: <input value="{{name}}"/>
ONE WAY c->p: 
ONE WAY c->v:
```

**Comment by justinbmeyer** [comment]
There's a problem with this and undefined values which is related to: #579.

Essentially, if `#foo` creates a `foo` value in a "references" scope, it will be undefined temporarily in an example like:

```html
<parent>
<child>
  {{foo}}
</child>
</parent>
<something #foo/>
```

In this case `{{foo}}` will be bound to chid.foo.

There are two fixes:

1.  Fix #579.


2.  Change `<my-comp #child/>` to `<my-comp @child/>` and create `@child` in the references scope.  Make values like `@NAME`, if they aren't found, default to observing on the references scope.


These a probably about the same difficulty, however, #579 might involve binding on every single "parent" observable context.  BOO!

I would prefer to write out references like `<custom #foo/>`, but `@foo` isn't so bad and wouldn't have potential performance issues.

**Comment by matthewp** [comment]
> Child to Parent	<my-comp [parent]="{childProp}" />

I think we should flip this around, it's weird that the direction is different between parent->child and child-> parent.  It's weird because the direction AND syntax are both different.  I think only the syntax should be different, I it makes it easier to understand.

#### child->parent

```html
<my-comp [child]="{parentProp"}>
```

corresponds to 

#### parent->child
```html
<my-comp child="{parentProp"}>
```

I almost wish we didn't have to distinguish between the direction and could have only 1 two-way binding syntax. What if you could do:

#### parent->child
```html
<my-comp [id]="{parentId}">
```

You currently can't do this with our binding syntax because `id` is reserved.

**Comment by justinbmeyer** [comment]
@matthewp I don't like the = sign going the other way.

**Comment by justinbmeyer** [comment]
I'm proposing changing this to account for problems in #1770.  A detailed explanation of the problems is discussed [here](https://gitter.im/bitovi/canjs?at=55f32a416fe7b2a123eb04d0).  


On a high level ... stache/handlebars/mustache should be its own language.  Bindings should be an interface that gets applied after stache has done its business.

On a lower level ... the original proposed binding syntaxes: `prop={}` and `prop={{}}` suffer from two problems:

1.  They make it difficult or impossible to use stache to "program" the bindings.  This is related to #1770.
2.  They use an attribute name they are adding behavior to.  For instance, `value="{key}"` would setup two-way binding on the `value` property.  Accounting for problem (#1), this means that the most straightforward implementation would have `value` set to "{key}" temporarily.

Finally, after [studying angular 2.0's syntax](http://victorsavkin.com/post/119943127151/angular-2-template-syntax), it is able to smarty combine event syntax `(event)` with one-way binding `[key]` to setup two-way binding like: `[(key)]`.  This is smart.  Lets steal it!

So, I'll be updating the proposal to closely match angular 2.0's syntax.

**Comment by justinbmeyer** [comment]
Here's the old proposal for safe keeping:

Two Way:

```
<component child-prop="{parentProp}">
```

One Way into child:

```
<component child-prop="{{parentProp}}">
<component child-prop="[parentProp]">
```

One way into parent:

```
<component [parent-prop]="{{childProp}}">
```

One way into variable

```
{{#}}
<component [parent-prop]="{{childProp}}">
{{/}}
```

**Comment by pYr0x** [comment]
whats the difference between `two way - viewModel` and `two way - reference` ?

if you bind a childProp to parentProp in the scope (two way - viewModel) thats the same as driver.plateName property to licensePlate (two way - reference)

**Comment by justinbmeyer** [comment]
The `reference` scope is a different scope that is local just to the current template.  This talks about it: http://canjs.com/2.3-pre/docs/can.view.bindings.reference.html

**Comment by matthewp** [comment]
```
<my-comp [(child-prop)]="#ref"/>
```

Why is the hash on the right side?

**Comment by justinbmeyer** [comment]
@matthewp So there, might be a need for parent after all.  Basically, I want a 1-way binding from a component property to `#ref`.  Something like:

```
<my-component ^child="#ref">
```

Boo!  This makes me think that two-way binding might look like:

```
<my-component [^child]="#ref">
```

vs

```
<my-component [(child)]="#ref">
```

Yarg.  I'm not sure what to do.  I thought about making 

```
<component [#ref]="foo"
```

Do one-way from #ref to foo.  However, I don't like that.  I think we should stick to the key always being the attribute value.

**Comment by justinbmeyer** [comment]
Alternatively, we could use `{child}="parent"` for child->parent one-way binding.  Making `[{child}]="parent"` also work for 2-way.

**Comment by justinbmeyer** [comment]
An alternate to the alternate would be to use one of the following:

| Type                 | Example  |
|-------------------------|--------------------------|
| one way - viewModel - parent to child | `<my-comp child-prop{="parentProp"/>` |
| one way - viewModel - child to parent | `<my-comp child-prop}="parentProp"/>` |
| two way - viewModel - child to parent | `<my-comp {child-prop}="parentProp"/>` |

| Type                 | Example  |
|-------------------------|--------------------------|
| one way - viewModel - parent to child | `<my-comp child-prop[="parentProp"/>` |
| one way - viewModel - child to parent | `<my-comp child-prop]="parentProp"/>` |
| two way - viewModel - child to parent | `<my-comp [child-prop]="parentProp"/>` |

**Comment by justinbmeyer** [comment]
Can't use & for reference because `{{&foo}}` is a mustache thing.  Damn.  Switching to `.

**Comment by justinbmeyer** [comment]
This almost makes me want to wait for 3.0 for references.  We can drop `{{&foo}}`.

**Comment by justinbmeyer** [comment]
Updated to what I think is the final syntax.

**Comment by matthewp** [comment]
Does this fix the Chrome warning about using `cloneNode` with Attributes?  There are two possible ways to fix this:

1. Create a new element and use innerHTML to set the attribute, just for the sake of getting it's Attribute node.  This is done [here](https://github.com/canjs/dom-patch/blob/master/src/setattribute.js#L23).
2. Rename these special attributes (which stinks to special case) into something like `one-way-parent-foo="bar"` and `dom-two-way-value="foo"`, these will be pretty long for some of the options.

**Comment by justinbmeyer** [comment]
@matthewp strangely I didn't notice an error.  Perhaps chrome removed it.  When you get latest minor, let me know if it happens to you.

**Comment by JustinLove** [comment]
Another way to do arrows, although I'm still not sold on arrows

```
child-prop{="}parentProp"
child-prop{="parentProp"
child-prop="}parentProp"
```

**Comment by JustinLove** [comment]
I've feel like I've got a solved on paper type solution to half the problem; limited by my rather recent approach to the problem. It has a quality of cohesion I've felt lacking elsewhere, likely biased by inventing it. No doubt some people will dislike this as much as I dislike the other options. ;^)

## Principles

### plain old html looks like plain old html

`prop=...` should be an dom property and `...="foo"` should be a plain string

### code looks like code

It would be nice if it was similar to `{{}}`, but doing that without colliding with the existing `{}` has proved daunting

Proposal: An attribute or value prefixed with `.` is read as a property reference into the implied context.  The implied context on the left is viewModel, and on the right the template context.

Special contexts can be used before the dot: `$` for the the dom element, and `*` for the reference scope.

### Clearly identify values being subscribed to

Proposal: `:` is an object reference "doubled-down" to a binding. `!` is a variant for events.

### Clearly identify values that can be changed externally

Proposal (not happy with part yet): I've tried to extend the expanded syntax, e.g. `~complete($element.checked)`, with an "implied value from other side" using `_`; However I feel like the postfix `(_)` is a bit verbose, and perhaps semantically unnecessary.

Type | Example
----- | -----
0 way - viewModel | `.child-prop="foo"`
one way - viewModel - parent to child | `.child-prop(_)=":parentProp"`
one way - viewModel - child to parent | `:child-prop=".parentProp(_)"`
two way - viewModel | `:child-prop(_)=":parentProp(_)"`
event - viewModel | `!child-event=".parentMethod()"`
event - DOM | `$!click=".parentMethod()"`
one way - DOM - parent to element | `$.el-prop(_)=":parentProp"`
two way - DOM | `$:el-prop(_)=":parentProp(_)"`
one way - reference - child to reference | `:child-prop="*.ref(_)"`
two way - reference | `:child-prop(_)="*:ref(_)"`
two way - reference shorthand | `:*:ref` or `*::ref`

**Comment by justinbmeyer** [comment]
@JustinLove I really like the idea that "." is reading from the element's viewmodel.  The rest of it I'm more ambivalent about.  Thanks for sharing!

Unfortunately however, this issue is closed (It was going figuratively until minor is in master).  I've implemented the proposal listed in the first issue and we've updated our DoneJS guides with this.  And I've begun the documentation. There's just too much behind it now to have another round of discussion.  

Hopefully what's prosed, if not ideal for everyone, is good enough for everyone. Thanks again!
